InteractiveRenderUI: use buttons to drive the state

mattigruener opened the pull request:
Undo currently potentially changes the state of the interactive rendering. This changes that by adding a plugValueWidget that drives that plug using a set of buttons.

@johnhaddon Not sure how you feel about some of the design decisions here. I tend to use namedtuples because it makes the code that uses them more readable. It's a bit overkill in this case. Let me know if there's anything you want me to change or approach completely differently. I added this to the base class because there was a todo tag that indicated that this was the place for it rather than just adding it to the Arnold node.

johnhaddon commented:
Thanks Matti!

I do agree that namedtuples can be a great way of making code more readable. In this case I think the extra level of indirection involved actually makes the code quite a bit harder to understand though. The tuple and the label map seem only to be used in `_updateButtons`, so by introducing them, the reader is forced to jump about the code rather than just read that single function. Not a biggy - I've got a few more important comments I'll make inline, so see how you feel about it when you're addressing those.

One other thing. In my head I had imagined buttons with icons - what do you think about that? I'm thinking I might promote this plug into the viewer toolbar for the new shader view, at which point I think icons might be a nice improvement. There are play/pause icons in resources/graphics.svg already, so we'd just need a stop I think.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+
+ButtonLabels = namedtuple( 'ButtonLabels', 'startButton pauseButton')
+
+class InteractiveRenderingStateWidget( GafferUI.PlugValueWidget ) :
```
Prefix with _, to make it clear that this is not part of the public API. Gaffer's codebase doesn't really follow the "we're all consenting adults" python approach of having everything hang out in public - things become unmaintainable really quickly if you don't know what client code might be using. In this case, it's hard to imagine client code accessing this particular class, but it's a principle worth applying everywhere consistently (and it's also sometimes scary what does end up getting used externally).

While we're here, _StatePlugValueWidget would probably be the more convention-following name.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+		GafferUI.PlugValueWidget.__init__( self, row, plug )
+
+		with row :
+			self.startButton = GafferUI.Button()
```
Prefix with `__`, to make it clear that this is an internal implementation detail. Same reason as above.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+			self.startButton = GafferUI.Button()
+			self.pauseButton = GafferUI.Button()
+
+			self.startConnection = self.startButton.clickedSignal().connect( Gaffer.WeakMethod( self._startStopCallback ) )
```
Gaffer has a convention of naming things with the following pattern :

- Signals : `clickedSignal()`
- Connections : `__clickedConnection()`
- Slots : `__clicked()`

This helps communicate the relationships between each of the things, and makes navigating the code a bit more predictable. So although it's wordy, I'd suggest `__startClickedConnection` and `__startClicked` (or something that reflects the dual start/stop nature of the button).

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+		self._updateButtons()
+
+	def _updateFromPlug( self ) :
+		pass
```
Bug alert! This should be the primary spot where we do the work of `_updateButtons()`. Without it, when there's more than one NodeEditor open for this node, the second one will not update when this one edits the value. Or if the plug has an input that provides the value, neither UI will ever update.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+	def _updateFromPlug( self ) :
+		pass
+
+	def hasLabel( self ) :
```
I'd imagined the label still being there, perhaps initially only because I'd also imagined icons and they'd be tiny without it. But I do think there might be valid workflows where you'd want to promote the state plug, in which case the label is necessary.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+	def hasLabel( self ) :
+		return True
+
+	def _isRunning( self ) :
```
This is just a taste thing perhaps, but I find having an `_isRunning()` function which just does `return state == Running` to be another example of indirection which makes things less readable.

johnhaddon replied:
Actually, I think it's more than a taste thing. The usage of these methods includes things like `self._isStopped() or self._isPaused()`, so the utility methods are designed such that two calls to `getValue()` are required where one would be sufficient. I think the following pattern is preferable (also taking into account my comment about contexts) :

```
with self.getContext() :
    state = self.getPlug().getValue()

if state == GafferScene.InteractiveRender.State.Stopped or state == GafferScene.InteractiveRender.State.Paused :
    ...

```

Now, I admit that those verbose accesses to the State enum are a drag, so I can understand the motivation to hide them somehow. But I think the code structure is more important in this case, and if you wanted to you could use an alias for `GafferScene.InteractiveRender.State` you could improve readability without changing the structure.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+	def _isPaused( self ) :
+		return self.plug.getValue() == GafferScene.InteractiveRender.State.Paused
+
+	def _startStopCallback( self, button ) :
```
Prefix all the internal methods with `__` (two instead of one).

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+			self.plug.setValue( GafferScene.InteractiveRender.State.Running )
+		else:
+			self.plug.setValue( GafferScene.InteractiveRender.State.Stopped )
+		self._updateButtons()
```
Once `_updateFromPlug()` is implemented correctly, this call will be unnecessary. In general, this "change the graph then update my ui" pattern is a bit of an indicator of bugs - the UI should always be updated via a signal of some sort, so that it can respond to externally triggered changes as well as internally triggered ones. The nice side effect of this is that you never need to explicitly make an update after triggering a graph change internally.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+
+	def _startStopCallback( self, button ) :
+		if self._isStopped() or self._isPaused() :
+			self.plug.setValue( GafferScene.InteractiveRender.State.Running )
```
Because not using an UndoContext here is out of the ordinary, I think it merits a comment explaining that undo is undesirable for the state plug, and has been omitted deliberately.

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+		return self.plug.getValue() == GafferScene.InteractiveRender.State.Running
+
+	def _isStopped( self ) :
+		return self.plug.getValue() == GafferScene.InteractiveRender.State.Stopped
```
Generally, accessing the value of a plug without using a context is an alarm bell - if the plug is driven by an expression that requires specific context variables from the script, everything fails. So all plug accesses need to follow this pattern :

```
with self.getContext() :
   value = self.getPlug().getValue()
```

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -36,6 +36,70 @@
+
+	def __init__( self, plug, *args, **kwargs) :
+
+		self.plug = plug
```
The plug is stored by the `PlugValueWidget` base class, and should be accessed via `self.getPlug()`. Storing it here is redundant, and also a source of bugs. There's a `setPlug()` method on the base class that , if called, would make `self.plug` not just redundant but also just wrong.

johnhaddon commented:
Blimey, I've made rather a lot of comments there haven't I? My intention isn't to discourage you - I think this was a good first stab. But I do see this as a good opportunity to introduce you to some of the conventions used in the Gaffer codebase, and also point out some of the common pitfalls in implementing new UI within it. Hopefully you've found it useful...
Cheers...
John

johnhaddon commented:
![](http://4.bp.blogspot.com/-B3-B9fpiJps/VlJOXn5pbJI/AAAAAAAABOM/NM8dx-uPYDM/s1600/columbo-quote-one-more-thing.jpg)

Am I right in thinking that it's more common for play/pause to share a button and for stop to be the separate one?

mattigruener commented:
Haha, thanks - not discouraged at all ;) I'm happy to learn the conventions.

Re play/pause + stop button: I pretty much replicated what we currently have in our in-house preview nodes. You're right, though. 

Re Icons: Sounds good. I didn't know you could promote these to the viewer directly.

I'll go through the rest of your notes hopefully today. Cheers!

mattigruener commented:
If my friday night brain isn't mistaken, I have addressed all your comments here. I also added the stop icon. I attach a screenshot of how the buttons currently look. Feel free to add more notes if anything is still not how it's supposed to be. Thanks! :)

![buttons](https://cloud.githubusercontent.com/assets/19398650/17827414/5a57c7e8-6631-11e6-9b2f-7609348f2967.png)

johnhaddon commented:
Nice one - thanks Matti. I've merged along with a couple of [tiny tweaks](https://github.com/ImageEngine/gaffer/commit/791beeb48fc8b6addb34dff8df5d124c85dffe50) that I didn't think it was worth bothering you with and that certainly weren't essential.
